File: 05_composition.rb
# frozen_string_literal: true

scream  = ->(str) { str.upcase }
exclaim = ->(str) { "#{str}!" }
repeat  = ->(str) { "#{str} #{str}" }

compose = ->(*fns) {
  ->(x) {
    fns.reverse_each.reduce(x) { |acc, fn| fn.(acc) }
  }
}

pipe = ->(*fns) { ->(x) { fns.reduce(x) { |acc, fn| fn.(acc) } } }

with_exuberance_compose = compose.(repeat, exclaim, scream)
with_exuberance_pipe = pipe.(scream, exclaim, repeat)

puts with_exuberance_compose.('Ruby FP is ok')
puts with_exuberance_pipe.('Ruby FP is ok')

File: 05_composition.js
const scream  = str => str.toUpperCase()
const exclaim = str => `${str}!`
const repeat  = str => `${str} ${str}`

const compose = (...fns) => x =>
  fns.reduceRight((acc, fn) => fn(acc), x)

const pipe = (...fns) => x =>
  fns.reduce((acc, fn) => fn(acc), x)

const withExuberance     = compose(repeat, exclaim, scream)
const withExuberancePipe = pipe(scream, exclaim, repeat)

console.log(withExuberance('JS FP is nice'))
console.log(withExuberancePipe('JS FP is nice'))
